engine/orchestration/src/com/cloud/vm/VirtualMachineManagerImpl.java

```diff
@@ -744,14 +744,17 @@ protected boolean checkWorkItems(final VMInstanceVO vm, final State state) throw
 
     protected <T extends VMInstanceVO> boolean changeState(final T vm, final Event event, final Long hostId, final ItWorkVO work, final Step step) throws NoTransitionException {
         // FIXME: We should do this better.
-        final Step previousStep = work.getStep();
-        _workDao.updateStep(work, step);
+        Step previousStep = null;
+        if (work != null) {
+            previousStep = work.getStep();
```

Comment:
I'm not sure I'm following.  From reading the code a bit, it looks like the only scenario where a false would be returned from stateTransitTo would be where the state was not properly persisted to the db.  No exception is thrown in that case.  In the current code, false *or* exception will try to revert.  Also, currently previousStep would *probably* never be null - though if work is null this will cause another NPE on line 747 of current code.  In the PR, previousStep will *probably* only be null in the case where work is null.

What is the value of making a distinction from a false versus an exception in this case?  Me adding a check for previousStep != null is simply to make sure that work is not null above.  I could also check work for null here too, but I don't think that's what you're getting at.